lib/transport/datagouvfr/client.ex

```diff
@@ -3,159 +3,61 @@ defmodule Transport.Datagouvfr.Client do
-      {:error, error} -> error
-    end
-  end
 
-  @doc """
-  Call to GET /api/1/organizations/:slug/
-  You can see documentation here: http://www.data.gouv.fr/fr/apidoc/#!/organizations/get_organization
-  """
-  @spec organizations(String.t) :: {atom, [map]}
-  def organizations(slug) do
-    case get(Path.join("organizations", slug),
-             [timeout: 50_000, recv_timeout: 50_000]) do
-      {:ok, %HTTPoison.Response{status_code: 200, body: body}} -> body
-      {:ok, %HTTPoison.Response{status_code: _, body: body}} -> {:error, body}
-      {:error, error} -> {:error, error}
-    end
-  end
-
-  @doc """
-  Call to GET /api/1/organizations/:slug/
-  And add datasets to it.
-  """
-  @spec organizations(String.t, atom) :: {atom, map}
-  def organizations(slug, :with_datasets) do
-    organizations(organizations(slug), slug, :with_datasets)
-  end
-
-  def organizations({:ok, body}, slug, :with_datasets) do
-    case datasets(%{:organization => slug}) do
-      {:ok, body_datasets} -> {:ok, Map.put(body, "datasets", body_datasets)}
-      {:error, error} -> {:error, error}
-    end
-  end
-
-  def organizations({:error, error}, _, _) do
-    {:error, error}
-  end
-
-  @doc """
-  Call to GET /api/1/organizations/:slug/datasets/
-  You can see documentation here: http://www.data.gouv.fr/fr/apidoc/#!/organizations/list_organization_datasets
-  """
-  @spec datasets(map) :: {atom, [map]}
-  def datasets(%{:organization => slug}) do
-    ["organizations", slug, "datasets"]
-    |> Path.join()
-    |> get([timeout: 50_000, recv_timeout: 50_000])
+  @spec get_request(%Plug.Conn{}, binary, OAuth2Client.headers, Keyword.t)
+                    :: {:ok, OAuth2.Response.t} | {:error, Error.t}
+  def get_request(%Plug.Conn{} = conn, url, headers \\ [], opts \\ []) do
+    :get
+    |> Transport.Datagouvfr.Client.request(conn, url, nil, headers, opts)
+  end
+
+  @spec put_request(%Plug.Conn{}, binary, OAuth2Client.body,
+                    OAuth2Client.headers, Keyword.t)
+                    :: {:ok, Response.t} | {:error, Error.t}
+  def put_request(%Plug.Conn{} = conn, url, body \\ "", headers \\ [], opts \\ []) do
+    headers = [{"content-type", "application/json"} | headers]
+    :put
+    |> Transport.Datagouvfr.Client.request(conn, url, body, headers, opts)
+  end
+
+  # We disable for now the credo test because the arity is to high
+  # TODO: add a Transport.Request module to lower the arity and make this
+  # module clearer
+  # credo:disable-for-lines:4
+  @spec request(atom, %Plug.Conn{}, binary, OAuth2Client.body,
+                OAuth2Client.headers, Keyword.t)
+                :: {:ok, Response.t} | {:error, Response.t} | {:error, Error.t}
+  def request(method, %Plug.Conn{} = conn, url, body, headers, opts) do
+    client = get_client(conn)
+    url = process_url(url)
+    method
+    |> Request.request(client, url, body, headers, opts)
+    |> post_process_request()
+  end
+
+  def get_client(conn) do
```

Comment:
IMHO this is a smell, as we're hiding an implementation problem upstream.

Reply:
what do you mean?

Reply:
with this, transport.Datagouvfr.Client is independent of the connection construction, it's easier to test.

Reply:
OK